### PR TITLE
Add new rule rsyslog_filecreatemode

### PR DESCRIPTION
**Pull request by dodys**
#### Description:

- This rule has been missing for a while and should solve issue #7332 
- This PR won't include ansible remediation, I expect other vendors will properly implement it.

#### Rationale:

- This rule is needed for CIS on Ubuntu, RHEL and SLES as far as I know, might be needed for other vendors as well.

### CODE REVIEW

File: linux_os/guide/system/logging/rsyslog_filecreatemode/tests/filecreatemode_0600.pass.sh [hunk 1]
```diff
@@ -0,0 +1,7 @@
+#!/bin/bash
+# platform = multi_platform_ubuntu
```
Comment (marcusburghardt) [inline linux_os/guide/system/logging/rsyslog_filecreatemode/tests/filecreatemode_0600.pass.sh:2]: I would suggest to remove `platform` restriction in these test scenarios since the tests seem reasonable regardless of product.

**Review by marcusburghardt — CHANGES_REQUESTED** [review]
Thanks for this rule @dodys . Please, take a look in my comments to make it simpler.

File: linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml [hunk 2]
```diff
@@ -0,0 +1,112 @@
+<def-group>
+  <definition class="compliance" id="{{{ rule_id }}}" version="1">
+    {{{ oval_metadata("rsyslog will create logfiles that do not already exist on the system. This setting controls what permissions will be applied to these newly created files.") }}}
+    <criteria operator="ONE">
```
[the remark this answers is not in the document]
Comment (dodys) [reply linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:4]: It was indeed chosen to be `ONE`, because it is not about the value of $FileCreateMode, but if it is defined in both places.
Let's say you have `$FileCreateMode 0600` in `/etc/rsyslog.conf` and you have `$FileCreateMode 0640` in `/etc/rsyslog.d/99-rsyslog_filecreatemode.conf`, both values are valid and will lead to both criterias being `True`. Then the `ONE` will make it fail as only one of the criterias should be `True`.

If you have `OR` in the case above, then it will pass when it shouldn't.

Therefore I believe the current check already covers your scenario.

[the remark this answers is not in the document]
Comment (dodys) [reply linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:4]: If we try to have the check you mentioned first, and then a variable for checking the value, I still think you will need two variables to convert from octal. Because you will have two tests, consequently two objects, to check `rsyslog.conf` and `rsyslog.d`. You will need a variable for each object, right?
Or do you have something else in mind?

[the remark this answers is not in the document]
Comment (marcusburghardt) [reply linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:4]: Thanks for the clarification @dodys . I forgot about the ONE value and didn't find another use in the project. But checked the [OVAL documentation](https://oval-community-guidelines.readthedocs.io/en/latest/oval-schema-documentation/oval-common-schema.html?#operatorenumeration-1) now and you are correct.

[the remark this answers is not in the document]
Comment (marcusburghardt) [reply linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:4]: > If we try to have the check you mentioned first, and then a variable for checking the value, I still think you will need two variables to convert from octal. Because you will have two tests, consequently two objects, to check `rsyslog.conf` and `rsyslog.d`. You will need a variable for each object, right? Or do you have something else in mind?

My point was more about the duplication of code and I was wondering if it could be somehow improved. Maybe a new OVAL macro would interesting since this conversion is used in at least 5 other rules. It would be great, but I would be fine leaving this out of scope in this PR. Maybe a separate PR just for the macro and the update of the impacted rules would be better.

Besides the macro, the other idea would demand to change the approach. For example:
First Test:
- Collect the FileCreateMode value in both `/etc/rsyslog.conf` and `rsyslog.d`.
  - It would still demand two objects, like now.
- Concatenate the objects in a new variable
  - It would demand only one conversion. We could consider only the first object, for example, in case the concatenated variable contains multiple objects.
- Compare the variable object to the expected state.

Second Test
- Ensure that concatenated variable contains only one object.

It would reduce the number of tests from 4 to 2 and would reduce code duplication.
Maybe it would be worth to consider.

File: linux_os/guide/system/logging/rsyslog_filecreatemode/tests/filecreatemode_0600.pass.sh (hunk 1 above)
Comment (dodys) [reply linux_os/guide/system/logging/rsyslog_filecreatemode/tests/filecreatemode_0600.pass.sh]: removed it from all tests

File: linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml (hunk 2 above)
[the remark this answers is not in the document]
Comment (dodys) [reply linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml]: I think I got it now, let me know what you think

File: linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml [hunk 3]
```diff
@@ -0,0 +1,112 @@
+<def-group>
+  <definition class="compliance" id="{{{ rule_id }}}" version="1">
+    {{{ oval_metadata("rsyslog will create logfiles that do not already exist on the system. This setting controls what permissions will be applied to these newly created files.") }}}
```
[the remark this answers is not in the document]
Comment (dodys) [reply linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml]: done, let me know what you think

File: linux_os/guide/system/logging/rsyslog_filecreatemode/rule.yml [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    It is important to ensure that log files have the correct permissions
+    to ensure that sensitive data is archived and protected.
+
+severity: medium
+
+references:
+    cis@ubuntu2004: 4.2.1.4
+    cis@ubuntu2204: 4.2.2.4
+
+ocil_clause: '$FileCreateMode is not set or more permissive than 0640'
```
Comment (marcusburghardt) [inline linux_os/guide/system/logging/rsyslog_filecreatemode/rule.yml:22]:
```suggestion
ocil_clause: '$FileCreateMode is not set or is more permissive than 0640'
```

File: linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml [hunk 5]
```diff
@@ -0,0 +1,62 @@
+<def-group>
+  <definition class="compliance" id="{{{ rule_id }}}" version="1">
+    {{{ oval_metadata("FileCreateMode setting controls permissions applied to newly created files.") }}}
+    <criteria operator="AND">
+      <criterion test_ref="tst_filecreatemode_declared"
+          comment="filecreatemode declared in either /etc/rsyslog.conf or /etc/rsyslog.d/*"/>
```
Comment (marcusburghardt) [inline linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:6]:
```suggestion
          comment="FileCreateMode declared once in either /etc/rsyslog.conf or /etc/rsyslog.d/*"/>
```

File: linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+      <criterion test_ref="tst_filecreatemode_valid"
+          comment="filecreatemode value is valid" />
```
Comment (marcusburghardt) [inline linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:8]:
```suggestion
          comment="FileCreateMode value is valid"/>
```

File: linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    </criteria>
+  </definition>
+
+  <ind:textfilecontent54_test id="tst_filecreatemode_declared"
+      version="1" comment="rsyslog filecreatemode configured"
```
Comment (marcusburghardt) [inline linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:13]:
```suggestion
      version="1" comment="rsyslog FileCreateMode is configured in only one place"
```

File: linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+      check="all" check_existence="only_one_exists">
+    <ind:object object_ref="obj_filecreatemode" />
+  </ind:textfilecontent54_test>
+
+  <ind:textfilecontent54_object id="obj_filecreatemode" version="1">
+    <ind:filepath operation="pattern match">\/etc\/rsyslog(\.conf|\.d\/.*\.conf)</ind:filepath>
```
Comment (marcusburghardt) [inline linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:19]:
```suggestion
    <ind:filepath operation="pattern match">\/etc\/rsyslog(\.conf|\.d\/.*\.conf)$</ind:filepath>
```

File: linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    <ind:pattern operation="pattern match">^\$FileCreateMode\s+(\d+)</ind:pattern>
+    <ind:instance datatype="int">1</ind:instance>
+  </ind:textfilecontent54_object>
+
+  <local_variable id="var_filecreatemode_dec"
+      comment="decimal conversion of octal filecreatemode"
```
Comment (marcusburghardt) [inline linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:25]:
```suggestion
      comment="decimal conversion of octal value from FileCreateMode parameter"
```

File: linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        </regex_capture>
+      </arithmetic>
+      <regex_capture pattern="\d\d\d(\d)">
+        <object_component object_ref="obj_filecreatemode" item_field="subexpression" />
+      </regex_capture>
+    </arithmetic>
+  </local_variable>
+
+  <ind:variable_test id="tst_filecreatemode_valid" version="1"
+      comment="Test if filecreatemode value is valid" check="all">
```
Comment (marcusburghardt) [inline linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:47]:
```suggestion
      comment="Test if FileCreateMode value is valid" check="all">
```

File: linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    <ind:object object_ref="obj_filecreatemode_dec" />
+    <ind:state state_ref="ste_correct_permissions" />
+  </ind:variable_test>
+
+  <ind:variable_object id="obj_filecreatemode_dec" version="1">
+    <ind:var_ref>var_filecreatemode_dec</ind:var_ref>
+  </ind:variable_object>
+
+  <!-- state to compare to -->
+  <ind:variable_state id="ste_correct_permissions" version="1">
```
Comment (marcusburghardt) [inline linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:57]: Could you rename this state for a name which suggest this rule, please? This is not strictly necessary, but it would reduce the chance of conflicts using more generic names and would also be easier to identify in a Data Stream.

File: linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+    <!-- MODE octal 640 converted to decimal: 6 * 64 + 4 * 8 + 0 = 416 -->
+    <ind:value operation="bitwise or" datatype="int">416</ind:value>
+  </ind:variable_state>
+
```
Comment (marcusburghardt) [inline linux_os/guide/system/logging/rsyslog_filecreatemode/oval/shared.xml:61]:
```suggestion
```

**Review by marcusburghardt — CHANGES_REQUESTED** [review]
Amazing @dodys . I really liked the new approach. Much simpler and easier to review. I have only minor comments and we should be ready. I also tested this rule in RHEL products and it is working fine. I will be happy to propose the Ansible remediation for it soon.

File: linux_os/guide/system/logging/rsyslog_filecreatemode/bash/shared.sh [hunk 13]
```diff
@@ -0,0 +1,5 @@
+# platform = multi_platform_all
+
+sed -i '/^\s*$FileCreateMode/d' /etc/rsyslog.conf /etc/rsyslog.d/*
```
Comment (marcusburghardt) [inline linux_os/guide/system/logging/rsyslog_filecreatemode/bash/shared.sh:3]: Thinking about this remediation, wouldn't it be better to first check if the value is correct and only change if necessary? If the admin decided to configure it correctly, but in a different file, by any reason, the remediation would unnecessarily change the file. This might be undesired in some cases.

Comment (dodys) [reply linux_os/guide/system/logging/rsyslog_filecreatemode/bash/shared.sh:3]: let me know if you have thoughts about it or if you did it differently for the ansible

Comment (marcusburghardt) [reply linux_os/guide/system/logging/rsyslog_filecreatemode/bash/shared.sh:3]: In Ansible, the logic would be something like:
- Find all rsyslog files containing $FileCreateMode
  - There is a rule using a similar approach:
  - https://github.com/ComplianceAsCode/content/blob/master/linux_os/guide/system/logging/rsyslog_sending_messages/rsyslog_remote_tls/rule.yml
- Concatenate these files in a list.
- If the list contains more than 1 file, remove the parameter in all files and create a new file specific for the parameter, in order to avoid conflicts.
- If the list contains only 1 file, ensure the value is correct on that file. Just update the value, if necessary.
- If the list is empty, create a new file with the parameter.

I believe a similar logic, maybe simpler, could be used in bash to make it more idempotent.
Basically, it would be necessary to only implement these tests in the script.

Comment (dodys) [reply linux_os/guide/system/logging/rsyslog_filecreatemode/bash/shared.sh]: sorry for the delay, I've just rebased the branch and improved the remediation.
Let me know if that looks good.

**Comment by marcusburghardt** [comment]
Automatus tests are failing because the rule is restricted to `ubuntu`. Waived!

**Comment by marcusburghardt** [comment]
Overriding CODEOWNERS since a @dodys can't approve his own PR.

**Comment by marcusburghardt** [comment]
My last tests finished and actually some two test scenarios are failing in RHEL7 only. I quickly checked and it seems to be a minor issue related to compatibility. Since this rule will be interesting for CIS in RHEL7 and I plan to include it pretty soon, I am fine to merge it now and fix any possible issue in RHEL7 when enabling this rule for RHEL.

**Comment by marcusburghardt** [comment]
> My last tests finished and actually some two test scenarios are failing in RHEL7 only. I quickly checked and it seems to be a minor issue related to compatibility. Since this rule will be interesting for CIS in RHEL7 and I plan to include it pretty soon, I am fine to merge it now and fix any possible issue in RHEL7 when enabling this rule for RHEL.

Fixed in #10328
